### ISSUE TITLE
nth is faster than nth-unsafe in some cases

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#1339]
I noticed that ``nth`` is faster than ``nth-unsafe`` for ``bit-arrays`` and other sequences.

```factor
: foo ( n -- m )
    [ 0 ] dip dup <bit-array>
    '[ _ nth [ 1 + ] when ] each-integer ;

IN: scratchpad gc [ 100,000,000 foo ] time .
Running time: 0.48131244 seconds
```

vs.

```factor
: bar ( n -- m )
    [ 0 ] dip dup <bit-array>
    '[ _ nth-unsafe [ 1 + ] when ] each-integer ;

IN: scratchpad gc [ 100,000,000 bar ] time .
Running time: 0.652339592 seconds
```

Seems to have something to do with the way ``bounds-check`` works inside of ``M\ sequences nth`` which allows some kind of fixnum type propagation of ``integer>fixnum`` that does not work inside of ``M\ bit-arrays nth-unsafe``.

**Comment by mrjbq7** [comment]
@bjourne or @ajvondrak might be interested in looking at this.

**Comment by ajvondrak** [comment]
> Seems to have something to do with the way `bounds-check` works inside of `M\ sequences nth` which allows some kind of fixnum type propagation of `integer>fixnum` that does not work inside of `M\ bit-arrays nth-unsafe`.

Yeah, looks like it.

For the benefit of the thread, the compiler.tree.debugger output:

```
IN: scratchpad \ foo optimized.
[
    >R 0 R> dup dup 0 < [ bad-array-length ] [ ] if dup 7 +
    -3 shift
    <byte-array> >R >fixnum 288230376151711743 fixnum-bitand R> 
    {
        bit-array
        2
        1
        tuple
        258304024774
        bit-array
        229512178467838274
    } <tuple-boa>
    "COMPLEX SHUFFLE" 0 "COMPLEX SHUFFLE" \ ( gensym ) [
        pick pick < [
            ( 7984990 7984994 7985760 -- 7984990 7985760 7984990 7984994 7985760 )
            >R >R >R 2dup 2 slot fixnum< [ \ t ] [ f ] if
            [ ] [ bounds-error ] if
            >R dup >R 7 >R R> fixnum-bitand
            R> -3 fixnum-shift-fast R> 3 slot
            swap alien-unsigned-1 swap >R R> -1 fixnum*fast
            fixnum-shift-fast 1 >R R> fixnum-bitand 0 eq?
            [ f ] [ \ t ] if [
                1 over tag 0 eq?
                [ fixnum+ ] [ fixnum>bignum bignum+ ] if
            ] [ ] if R> R> R> "COMPLEX SHUFFLE" 1 fixnum+fast
            "COMPLEX SHUFFLE" ( gensym )
        ] [ 3drop ] if
    ] label
]
```

```
IN: scratchpad \ bar optimized.
[
    >R 0 R> dup dup 0 < [ bad-array-length ] [ ] if 7 + -3 shift
    <byte-array> >R R> "COMPLEX SHUFFLE" 0 "COMPLEX SHUFFLE"
    \ ( gensym ) [
        pick pick < [
            ( 7985930 7985934 7986573 -- 7985930 7986573 7985930 7985934 7986573 )
            >R >R >R >R integer>fixnum dup >R 7 >R R>
            fixnum-bitand R> -3 fixnum-shift-fast
            R> swap alien-unsigned-1 swap >R R> -1 fixnum*fast
            fixnum-shift-fast 1 >R R> fixnum-bitand 0 eq?
            [ f ] [ \ t ] if [
                1 over tag 0 eq?
                [ fixnum+ ] [ fixnum>bignum bignum+ ] if
            ] [ ] if R> R> R> "COMPLEX SHUFFLE" 1 over tag 0 eq?
            [ fixnum+ ] [ fixnum>bignum bignum+ ] if
            "COMPLEX SHUFFLE" ( gensym )
        ] [ 3drop ] if
    ] label
]
```

With enough squinting, we see the main differences between the two are:
* The loop in `foo` does a `1 fixnum+fast` to increment the loop counter, whereas `bar` has a type check: `1 over tag 0 eq? [ fixnum+ ] [ fixnum>bignum bignum+ ] if`
* The bit-array tuple is constructed in `foo`, whereas `bar` has the underlying byte-array inlined
* I think the most crucial clue is this odd `288230376151711743 fixnum-bitand R>` in `foo`, which basically chops off the bit-array's `length` slot at a maximum of value of 288230376151711743; `bar` does no such length-limiting

My hypothesis is that the `bounds-check` propagates *interval* information about the bit-array's maximum length to the constructor, since 288230376151711743 is the maximum capacity of an array:

```
IN: scratchpad array-capacity-interval .
T{ interval { from { 0 t } } { to { 288230376151711743 t } } }
```

Because this gets propagated to the `<bit-array>` constructor, it knows its length slot (the `n` input to the `foo` word) is at most 288230376151711743, and makes sure of it with the `bitand`. Thus, interval/type propagation notices that the induction variable can fit in a fixnum:

```
IN: scratchpad fixnum-interval .
T{ interval
    { from { -576460752303423488 t } }
    { to { 576460752303423487 t } }
}
```

Therefore, the fixnum type information gets back through to the inside of the loop, where the type dispatch can be eliminated.

Since `nth-unsafe` doesn't do the `bounds-check`, we never limit the length of the underlying `<byte-array>` and consequently never propagate the type info into the loop.

I'm not entirely clear on how `bounds-check` winds up causing array capacity information to spread (probably because it calls `length`?), but that interval is certainly [propagated around](https://github.com/slavapestov/factor/blob/master/basis/compiler/tree/propagation/recursive/recursive.factor#L22). So I'm willing to bet it's the first domino to fall in the chain reaction that eliminates the type dispatch.

**Comment by bjourne** [comment]
Yeah, it's definitely the `bounds-check` that limits the range of the counter to [-inf,array-capacity]. And we already knew that was a fixnum so adding 1 to it is another fixnum so `fixnum+fast` is appropriate. Here is the same problem with regular arrays and recursion:
```factor
: (foo3) ( seq sum n len -- sum' )
    2dup < [
        ! seq sum n len
        [
            ! seq sum n | len
            pick dupd
            ! seq sum n n seq | len
            2dup 1 slot { array-capacity } declare < [ bounds-error ] unless
            nth-unsafe
            ! seq sum n el | len
            rot +
            ! seq n sum' | len
            swap 1 +
        ] dip
        ! seq sum' n len
        (foo3)
    ] [ 2drop nip ] if ; inline recursive

: foo3 ( n -- m )
    dup 3 <array> 0 rot 0 swap (foo3) ; inline
```
The bounds check is the line `2dup 1 slot { array-capacity } ...` and if you remove it you get slow code. I think the problem can be fixed by making the `nth-unsafe` method look like:
```factor
M: array nth-unsafe [ integer>fixnum { array-capacity } declare ] dip array-nth ; inline
```
Essentially adding the same type constraint `bounds-check` does. An even better but harder way would be to make sequence constructors propagate back their maximum ranges. E.g
```factor
[ dup 3 <array> drop ] final-info .
V{
    T{ value-info-state
        { class integer }
        { interval full-interval }
    }
}
```
It says `full-interval`, but it should instead be limited to `[0,max-array-capacity]` which is much smaller.

**Comment by bjourne** [comment]
Btw this is a place where Joe Groff's idea of forcing integers to fit into a fixnum interval to be fixnums would shine #224. Then since the stack parameter to `foo3` is used to construct an array, it must be a fixnum in the interval `[0,max-array-capacity]` so the loop would be:
```factor
\ foo3 optimized.
[
    dup 3 >R R> <array> 0 rot 0 swap \ ( gensym ) [
        2dup fixnum< [
            >R pick dupd >R R> swap 2 fixnum+fast slot rot +
            swap 1 fixnum+fast R> ( gensym )
        ] [ 2drop nip ] if
    ] label
]
```
And very good assembly code. Without it, you need to safeguard for bignums each loop iteration:
```factor
\ foo3 optimized.
[
    dup 3 <array> 0 rot 0 swap \ ( gensym ) [
        2dup dup tag 0 eq?
        [ fixnum< ] [ >R fixnum>bignum R> bignum< ] if [
            >R pick dupd >R R> swap 2 fixnum+fast slot rot +
            swap 1 fixnum+fast R> ( gensym )
        ] [ 2drop nip ] if
    ] label
]
```

**Comment by mrjbq7** [comment]
I think after 0.98 we should do this and #788 and related speedups.  Factor will be really awesome then.

**Comment by bjourne** [comment]
It's unfortunate that `array-capacity-max` is so small on 32 bit. it only goes to about 67 million and you are likely to sometimes want sequences larger than that. On 64 bit, the max is 288 230 376 151 711 743 and I don't think it would be a problem to limit the length to that. Even 32 bit `most-positive-fixnum` is to small at 134 million.

Like if you could assume that a sequences length fits in a fixnum, then the compiler could optimize like this:
```factor
[ [ . ] each ] optimized.
[
    dup >R length integer>fixnum R> "COMPLEX SHUFFLE" 0
    "COMPLEX SHUFFLE" \ ( gensym ) [
        pick pick fixnum< [
            ( 8968980 8968984 8969182 -- 8968980 8969182 8968980 8968984 8969182 )
            >R >R >R nth-unsafe .
            R> R> R> "COMPLEX SHUFFLE" 1 fixnum+fast
            "COMPLEX SHUFFLE" ( gensym )
        ] [ 3drop ] if
    ] label
]
```
It runs `1 fixnum+fast` to increment the counter variable without overflow checks every loop which is very good. But if you do not make that assumption, you get:
```factor
[ [ . ] each ] optimized.
[
    dup >R length dup integer?
    [ ] [ integer-length-expected ] if
    R> "COMPLEX SHUFFLE" 0 "COMPLEX SHUFFLE" \ ( gensym ) [
        pick pick 2dup both-fixnums?
        [ fixnum< ] [
            dup tag 0 eq?
            [ fixnum>bignum bignum< ] [
                over tag 0 eq?
                [ >R fixnum>bignum R> bignum< ] [ bignum< ] if
            ] if
        ] if [
            ( 8601560 8601564 8601764 -- 8601560 8601764 8601560 8601564 8601764 )
            >R >R >R nth-unsafe .
            R> R> R> "COMPLEX SHUFFLE" 1 over tag 0 eq?
            [ fixnum+ ] [ fixnum>bignum bignum+ ] if
            "COMPLEX SHUFFLE" ( gensym )
        ] [ 3drop ] if
    ] label
]```
The speed difference is pretty big too.

**Comment by mrjbq7** [comment]
I'd love to get word-size fixnum's somehow, or maybe use a single bit to tag fixnums and then if not a fixnum other tag bits to determine which type it is so you only lose one bit for fixnums.

**Comment by jonenst** [comment]
Is the problem slower iteration on other types of sequences, or that when there is no type propagation then the iteration is slower even for arrays ?

**Comment by bjourne** [comment]
Iteration is not slow for arrays, try for example `[ { array } declare [ sq ] map ] optimized.`. The reason is because we know that an array can't contain more than `max-array-capacity` elements and that constant is always smaller than `most-positive-fixnum`. That means the induction variable can fit in a fixnum.

Iteration is slow for arbitrary sequence types, see `[ [ sq ] map ] optimized.`. The reason is that they can contain more than `most-positive-fixnum` elements which means that the induction variable does not (necessarily) fit in a fixnum. So for each loop you need to check for overflow.

**Comment by jonenst** [comment]
What I meant by "when there is no type propagation" is for example what happens with lists:sequence>list. It is slow on arrays because it's not inlined. Is it a problem ?

I was asking this because it seems that for other sequences, especially virtual sequences, the "+1" with checks and bignum might be negligible compared to the rest of operations that are need for the iteration ?

Even so, most of the time, the function you are applying should take most of the time anyway ? like in the following, with arrays it's not so different when you map ```number>string``` :
```factor
IN: scratchpad : test-copy ( seq -- seq' ) [ ] map ;
IN: scratchpad : test-copy-fast ( seq -- seq' ) { array } declare [ ] map ;
IN: scratchpad : test-op ( seq -- seq' ) [ number>string ] map ;
IN: scratchpad : test-op-fast ( seq -- seq' ) { array } declare [ number>string ] map ;

IN: scratchpad 10000000 iota >array gc {
[ [ test-copy drop ] time ]
[ [ test-copy-fast drop ] time ]
[ [ test-op drop ] time ]
[ [ test-op-fast drop ] time ]
} cleave
```

Running time: 0.16341352 seconds
Running time: 0.036070578 seconds

Running time: 2.350969351 seconds
Running time: 2.026684119 seconds

**Comment by mrjbq7** [comment]
@jonenst it's true that one response is "does it matter that we are fast?", but I think it does.

See these issues for example: #1213, #878, #788.

The speed gains by "fixing" things like this are 10-50%.  That's a lot more than a "little difference".

**Comment by bjourne** [comment]
Yes I agree. @jonenst, for example you can change the `((each))` word in sequences:

```factor
: ((each)) ( seq -- n quot )
    [ length check-length ] keep [ nth-unsafe ] curry ; inline
```
to
```factor
: ((each)) ( seq -- n quot )
    [ length check-length integer>fixnum ] keep [ nth-unsafe ] curry ; inline
```
Then recompile everything and make a new bootstrap. On my computer, bootstrap times drop from 6m20s to 5m20s which is a lot for a small change.

**Comment by jonenst** [comment]
I wonder what kind of sequences are being iterated on during bootstrap. I would think mostly arrays ? So maybe the problem is what I said earlier:
> What I meant by "when there is no type propagation" is for example what happens with lists:sequence>list. It is slow on arrays because it's not inlined. Is it a problem ?

Also, maybe the minute difference is only due to improving the non optimized part of the bootstrap ?

**Comment by mrjbq7** [comment]
But it's not just unoptimized words that are slow... two versions, one that doesn't know its type and one that does know it's an array:

```factor
: simple-sum ( seq -- n )
   [ 0 ] dip [ + ] each ;

: array-sum ( seq -- n )
   { array } declare [ 0 ] dip [ + ] each ;
```

Have vastly different performances:

```factor
IN: scratchpad 100,000,000 0 <array> gc [ simple-sum drop ] time
Running time: 0.921614107 seconds

IN: scratchpad 100,000,000 0 <array> gc [ array-sum drop ] time
Running time: 0.376873728 seconds
```

**Comment by mrjbq7** [comment]
Inlining shouldn't be the answer, since if you inline everything code size explodes (which I ran into when I tried to fix #788).

**Comment by jonenst** [comment]
Bjourne maid a proposal:
> Like if you could assume that a sequences length fits in a fixnum

I said:
> Is the problem slower iteration on other types of sequences, or that when there is no type propagation then the iteration is slower even for arrays ?
> lists:sequence>list is slow on arrays because it's not inlined. Is it a problem ?

My point was: is bjourne's proposal 
1 ) because we want to iterate fast on all sequences ?
2 ) because we want to iterate fast on arrays even when there is no type propagation ?

**Comment by bjourne** [comment]
I think the commits I pushed could fix this problem. But they crash the build bots because they require new boot images.

**Comment by mrjbq7** [comment]
Do you have the password the run?

./factor -run=bootstrap.image.upload

**Comment by mrjbq7** [comment]
I can upload in an hour or so

**Comment by erg** [comment]
I uploaded!

On Fri, Mar 18, 2016 at 3:50 PM, John Benediktsson <notifications@github.com

**Comment by jonenst** [comment]
Could you summarize the changes ? I looked at the code but didn't get the big picture.
So in init-value-info, array-capacity is now replaced with fixnum. That's new right ? what does it do ?
And the new integer-array-capacity, which has the same min/max/interval as array-capacity is replaced with integer. Why do we need a new integer-array-capacity?

**Comment by mrjbq7** [comment]
@bjourne to get initial values to work (@jonenst comment on #1571)

```diff
diff --git a/core/slots/slots.factor b/core/slots/slots.factor
index e3f6ea9..6d393f1 100644
--- a/core/slots/slots.factor
+++ b/core/slots/slots.factor
@@ -203,6 +203,7 @@ M: anonymous-intersection initial-value*
         { [ dup "initial-value" word-prop ] [ dup "initial-value" word-prop t ] }
         { [ \ f bootstrap-word over class<= ] [ f t ] }
         { [ \ array-capacity bootstrap-word over class<= ] [ 0 t ] }
+        { [ \ integer-array-capacity bootstrap-word over class<= ] [ 0 t ] }
         { [ bignum bootstrap-word over class<= ] [ 0 >bignum t ] }
         { [ float bootstrap-word over class<= ] [ 0.0 t ] }
         { [ string bootstrap-word over class<= ] [ "" t ] }
```

I'm not going to patch the ``<string>`` or the ``{ integer-array-capacity } declare`` yet because I thought you should look at it and concur.

**Comment by bjourne** [comment]
That is probably right. I didn't intend for the class to be used in slot declarations, but it makes sense. I'll make a summary of how the propagation works in a little while.

**Comment by bjourne** [comment]
The type propagation system decorates each value with a class and an interval which it calls a "value-info". When the values flow through the code, they are constrained to smaller and smaller classes and intervals. Which is good, because smaller classes allows us to generate more efficient code. For example, + on a `number` is very expensive, + on an `integer` cheaper and + on `fixnum` cheaper still.

> So in init-value-info, array-capacity is now replaced with fixnum. That's new right ? what does it do ?

That is just normalization. Predicate classes in general are not orderable so `array-capacity integer-array-capacity class<=` is `f` even though all fixnums [0,max-array-capacity] is obviously less than all integers [0,max-array-capacity]. But if you replace `array-capacity` with the fixnum class and the given interval, the logic for comparing value infos will work out:

```factor
array-capacity <class-info> integer-array-capacity <class-info> value-info<=
t
```
Which it didn't before.

> Why do we need a new integer-array-capacity?

The more precise constraints we can place on values, the more efficient code Factor can generate. Unfortunately (from a performance perspective) we also want to be seamless between bignums and fixnums. `most-positive-fixnum 1 +` doesn't wrap around like some lame languages do. What that means is for the following:

```factor
[ { fixnum } declare 1 + ] optimized.
[ 1 fixnum+ ]
```
The most optimal code isn't generated because we have to take care of the overflow check. However, if you combine it with the `integer-array-capacity` declaration you get the optimal code:
```factor
[ { fixnum } declare { integer-array-capacity } declare 1 + ] optimized.
[ 1 fixnum+fast ]
```

**Comment by mrjbq7** [comment]
Are there any drawbacks to doing the ``fix-capacity-class``?  Is this just about the class of the variable, but you preserve the interval in the ``value-info``, so it's sort of like "unpacking"?

**Comment by bjourne** [comment]
Yes, that is kind of what it is. :) Before when you wrote `{ array-capacity } declare` you told the type system "this value is of class array-capacity" now it means "this value is of class fixnum AND in the interval [0,max-array-capacity]" The latter is more flexible. 

Without the change this expression is not flushed:
```factor
[ 20 f <array> drop ] optimized.
```
Which happens because `20 <literal-info> integer-array-capacity <class-info> value-info<=` is false but should be true. And I can't make that work without the "unpacking."
